scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+"""
+gensa: A generalized simulated annealing global optimization algorithm
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+import time
+from scipy.optimize import OptimizeResult
+from scipy.optimize import _lbfgsb
+from scipy.special import gammaln
+
+__all__ = ['gensa']
+
+KSPRING = 1.e8
+BIG_VALUE = 1.e13
+MAX_REINIT_COUNT = 100
+
+
+class GenSARunnerException(Exception):
+    pass
+
+
+class GenSARunner(object):
+    """This class implements the core of the gensa algorithm.
+
+    x0 : ndarray
+        The starting coordinates.
+    fun : callable
+        The objective function
+    lower: ndarray
```

Comment:
Consider having a single `bounds` variable, accepting `(min, max)` pairs. I.e. `gensa` just passes its bounds unchanged and all the bounds checking is done here.